web/src/app/shared/configuration/tools/filter/filter.component.ts

```diff
@@ -0,0 +1,170 @@
+              <mat-select placeholder="{{ filter.display_name }}" formControlName="{{ filter.filter_field }}"
+                          (selectionChange)="applyFilters()">
+                <mat-option *ngFor="let p of filter.options" [value]="p.value">{{ p.display_name }}</mat-option>
+              </mat-select>
+              <button mat-button matSuffix mat-icon-button aria-label="Clear"
+                      *ngIf="clearButtonVisible(filter.filter_field)"
+                      (click)="clear(filter.filter_field, $event)">
+                <mat-icon>refresh</mat-icon>
+              </button>
+              <button mat-button matSuffix mat-icon-button aria-label="Remove"
+                      (click)="removeFilter(filter, $event)">
+                <mat-icon>close</mat-icon>
+              </button>
+            </mat-form-field>
+          </ng-container>
+        </ng-container>
+      </form>
+    </div>
+  `,
+  styleUrls: ['./filter.component.scss'],
+})
+export class FilterComponent extends BaseDirective implements OnInit, OnDestroy {
+  filterForm = new FormGroup({});
+  availableFilters: any[];
+  activeFilters: number[] = [];
+  backupData: any;
+  freezeBackupData: boolean = false;
+  externalChanges: boolean = false;
+  @Input() filterList: IFilter[] = [];
+  @Input() externalData: BehaviorSubject<any>;
+  @Input() innerData: BehaviorSubject<any>;
+
+  get filters() {
+    return this.filterList.filter((filter) => (this.activeFilters?.includes(filter.id)));
+  }
+
+  constructor() {
+    super();
+  }
+
+  ngOnInit() {
+    this.availableFilters = this.filterList.map((filter: IFilter) => ({
+      id: filter.id,
+      name: filter.name,
+      display_name: filter.display_name,
+      filter_field: filter.filter_field,
+    }));
+
+    this.externalData.subscribe((values: any) => {
+      this.externalChanges = true;
+      this.freezeBackupData = false;
+
+      if (this.externalChanges && values) {
+        this.innerData.next(values);
+
+        this.innerData.subscribe((values: any) => {
+          if (!this.backupData || !this.freezeBackupData) {
+            this.backupData = values;
+            this.freezeBackupData = false;
+          }
+
+          if (this.externalChanges) {
+            this.externalChanges = false;
+            this.applyFilters();
+          }
+        })
+      }
+    });
+  }
+
+  clear(filter, event: any) {
+    this.filterForm.get(filter).setValue(undefined);
+    this.innerData.next(this.backupData);
+    event.preventDefault();
+    event.stopPropagation();
+  }
+
+  removeFilter(filter, event) {
+    this.toggleFilters(filter);
+    this.applyFilters();
+    event.preventDefault();
+  }
+
+  applyFilters() {
+    const filters = this.filterForm.value;
+    Object.keys(filters).forEach((f) => {
+      if (filters[f] === '' || filters[f] === undefined) {
+        delete filters[f];
+      }
+    });
+    const data = this.backupData?.results?.filter((item) => {
```

Comment:
in first view, want to filter this.externalData. What's reason for use this.backupData ?